### ISSUE TITLE
Intermediate decoder-block hidden states read back as all-inf while the final output is correct (Gemma-4-12B, compiled path)

### ISSUE THREAD

**Issue by Gal-bloch** [torch-spyre/torch-spyre#3240]
### What happened?

On the compiled Spyre backend, the model runs correctly end to end — greedy generation
on `google/gemma-4-12B-it` (fp16) **matches the CPU reference token-for-token**, and the
final logits are finite (softcapped at 30). So the per-layer residual/hidden states are
correct in the live block→block dataflow.

But we **cannot read those intermediate hidden states out of the compiled forward.** When
we tap a decoder block's output (to feed a drafter for EAGLE / speculative decoding), the
**first block reads back correctly, but every deeper block reads back as all-`inf`** —
via a raw host copy (`h.to("cpu")`), an on-device reduction (`h.sum()`), or an on-device
consumer (`h @ w`). All three ways of accessing the tapped value give `inf`, while the
same value flowing untouched to the next compiled block is fine (generation is correct).

In other words: the intermediate is valid *inside* the fused compiled block chain, but any
access to it from outside that fused dataflow returns garbage. This blocks reading
per-layer activations (an `output_hidden_states`-style capability), which EAGLE-style
speculative decoding requires.

```
generation:  matches CPU token-for-token          (live dataflow is correct)
logits:      inf=0     absmax=30.00                (final output valid)
block[0] out: finite   (sum=439, amax=15.5)        (first block taps correctly)
block[1] out: inf      (sum/amax/consumer/copyout all inf)
block[17] out: inf     (all inf)
block[46] out: inf     (all inf)
```

### What did you expect to happen?

Reading a decoder block's output hidden state should return the same finite values the
model uses internally (matching a CPU reference within fp16 tolerance) for every block,
not just block 0 — since the model demonstrably computes them correctly (generation
matches CPU). Equivalently: it should be possible to obtain per-layer hidden states from a
compiled forward for downstream use.

### How can we reproduce it?

`google/gemma-4-12B-it`, fp16, single prefill. Tap each compiled decoder block's output at
its boundary and read it three ways (host copy-out, on-device reduction, on-device
consumer matmul); the model's final logits confirm the forward itself is correct.

```python
import torch, torch_spyre  # noqa: F401
from transformers import AutoModelForCausalLM, AutoTokenizer
from hf_adapters import hf_common
from hf_adapters.auto_spyre_model import _resolve_adapter_module

MODEL = "google/gemma-4-12B-it"
CAPTURE = [0, 1, 17, 46]
dev = torch.device("spyre:0"); torch.spyre.set_device(0)
module = _resolve_adapter_module(MODEL)
tok = AutoTokenizer.from_pretrained(MODEL)
ids = tok("What are the main businesses of IBM? Answer in one sentence.",
          return_tensors="pt").input_ids
S = ids.shape[1]; pos = torch.arange(S).unsqueeze(0)

model = hf_common.load_model_common(MODEL, module.prepare_for_spyre, torch.float16,
                                    auto_model_cls=AutoModelForCausalLM)
D = int(getattr(model.config, "text_config", model.config).hidden_size)
w = (torch.randn(D, D, dtype=torch.float16) * 0.02).to(dev)   # on-device consumer

stats = {}
blocks = model._spyre_compiled_blocks
def wrap(i, fn):
    def wr(*a, **k):
        out = fn(*a, **k); h = out[0] if isinstance(out, tuple) else out
        if i in CAPTURE:
            s   = h.sum().to("cpu").item()                       # on-device reduction
            amx = h.abs().amax().to("cpu").item()
            con = (h.reshape(-1, D) @ w).sum().to("cpu").item()  # on-device consumer
            raw = int(torch.isinf(h.to("cpu").float()).sum())    # host copy-out
            stats[i] = (s, amx, con, raw, h.numel())
        return out                                               # untouched value flows on
    return wr
model._spyre_compiled_blocks = [wrap(i, b) for i, b in enumerate(blocks)]

kc, vc = hf_common.allocate_kv_caches(model, 1, S, torch.float16, device=dev)
mask = hf_common.build_prefill_mask(1, S, S, prompt_offsets=0, dtype=torch.float16).to(dev)
logits = module._run_forward(model, ids.to(dev), pos.to(dev), mask, kc, vc,
                             is_filling=False, token_index=0, cache_position=0)
lc = logits.to("cpu").float()
print("logits inf=", int(torch.isinf(lc).sum()), "absmax=", float(lc[torch.isfinite(lc)].abs().max()))
for i in CAPTURE:
    s, amx, con, raw, n = stats[i]
    print(f"block[{i}] sum={s:.3e} amax={amx:.3e} consumer={con:.3e} copyout_inf={raw}/{n}")
```

Expected on Spyre: `logits` finite (`absmax=30`), and greedy generation from this model
matches CPU; but `block[0]` taps finite while `block[1]/[17]/[46]` all read `inf`.

### Any environmental details we need to know?

- torch `2.11.0`
- torch-spyre @ `c9d435f`
- Model: `google/gemma-4-12B-it`, fp16, AIU 1.0 (DD2), single card, single prefill.
- Driven via the hf-adapters compiled path; hf-adapters only composes standard torch ops
  (a `torch.compile`'d per-layer block of matmul / SDPA / RMSNorm + residual) and loops
  them. The wrapper above only *observes* each block's returned tensor and passes it
  through unchanged, so the model math is unaffected (generation still matches CPU).

### Anything else we need to know?

- Proof the intermediates are correct on-device: greedy generation matches CPU
  token-for-token and the final logits are finite — the deep-layer residuals the model
  consumes are fine. Only *reading them out of band* fails.
- Block 0 (whose input is the embedding) reads correctly; every deeper block does not.
- The failing read is not specific to host transfer: an on-device reduction and an
  on-device consumer matmul of the tapped tensor also return `inf`, i.e. the tapped
  handle itself is invalid, not just its host copy.
- Minimal synthetic graphs (pure pointwise chains, standalone SDPA, a small
  norm-terminated attention block) do **not** reproduce it — their intermediates tap
  correctly — so it appears tied to the deep, fused compiled decoder-block chain.

### Relevant log output

```Shell
generation: matches CPU (token-for-token)
logits:     inf=0      absmax=30.00
block[0]:   sum=4.395e+02 amax=1.555e+01  consumer=-7.088e+01  copyout_inf=0/49920
block[1]:   sum=inf       amax=inf        consumer=inf         copyout_inf=49920/49920
block[17]:  sum=inf       amax=inf        consumer=inf         copyout_inf=49920/49920
block[46]:  sum=inf       amax=inf        consumer=inf         copyout_inf=49920/49920
```
